### ISSUE TITLE
Shared property instance when configuration cache enabled

### ISSUE THREAD

**Issue by yujincheng08** [gradle/gradle#24043]
Consider the following code snippet:

```kotlin
val seed = objects.property(Int::class.java)

val task1 = tasks.register("task1") {
    outputs.upToDateWhen {
        false
    }
    val s = seed
    doLast {
        val rand = SecureRandom().nextInt()
        println("set seed to $rand")
        s.set(rand)
    }
}

tasks.register("task2") {
    dependsOn(task1)
    val s = seed
    doLast {
        println("seed is ${s.get()}")
    }
}

tasks.register("task3") {
    dependsOn(task1)
    val s = seed
    doLast {
        println("seed is ${s.get()}")
    }
}
```
This code snippet tries to generate a random seed during the execution phase but not affecting the configuration phase.

### Expected Behavior
Task 2 or 3 should produce the same output without exception whether the configuration cache is enabled. 

### Current Behavior
When the configuration cache is enabled, tasks 2 or 3 will fail because the seed is not set.

### Context
Sometimes I want to make my build stochastic (e.g., integrating build time information). Because of the Gradle daemon, those stochastic contents should be generated in the execution phase and run only once. It's natural to create a task to modify a corresponding property, which dependent tasks can use then. However, when the configuration cache is enabled, the "value" of the property during the configuration phase is cached. Each task thus has a different instance of property (rather than sharing one). I think the configuration cache should be able to know and cache that some tasks "share" the same property.

I know `ValueSource`, and it works for stuff that does not change frequently. But it's unsuitable for stochastic contents that change every time (that always invalidate the configuration cache).

**Comment by yujincheng08** [comment]
This is a bit tricky because the property value may have changed when configuring tasks 2 and 3, and the configuration cache needs to cache this kind of change and still share the same property.

I think something like ExecutionPhaseOnlyProperty is helpful, or something like #22494 should also be okay (that does not invalid all configuration cache because one `ValueSource` changes)

**Comment by eskatos** [comment]
Thank you for your interest in Gradle!

Your issue appears to be a question about how to use Gradle. Your question might be answered in Gradle documentation or on our community Slack or forums. Find links to these resources on [help.gradle.org](https://help.gradle.org/).

If you think you have come across a Gradle bug, please file a new issue describing the problem and how to reproduce it.  

If you think Gradle is missing a feature, please file a new issue describing your use case for the feature.

***

Shared Build Services are the construct to use to share mutable state between tasks: https://docs.gradle.org/current/userguide/build_services.html

**Comment by bamboo** [comment]
For now this use-case has to be modeled using task outputs and inputs:
```kotlin
import java.security.SecureRandom

abstract class Seeder : DefaultTask() {

    @get:OutputFile
    abstract val outputFile: RegularFileProperty

    init {
        outputs.upToDateWhen {
            false
        }
    }

    @TaskAction
    fun seed() {
        val seed = SecureRandom().nextInt()
        outputFile.get().asFile.writeText(seed.toString())
    }
}

val task1 = tasks.register<Seeder>("task1") {
    outputFile.set(layout.buildDirectory.file("seed"))
}

val seedFile = task1.flatMap { it.outputFile }

(2..3).forEach {
    tasks.register("task$it") {
        // this should be a proper task @InputFile property
        inputs.file(seedFile)
        doLast {
            val seed = inputs.files.singleFile.readText().toInt()
            println("seed is $seed ")
        }
    }
}
```
